datafusion/substrait/tests/cases/roundtrip_logical_plan.rs

```diff
@@ -977,6 +984,43 @@ async fn new_test_grammar() -> Result<()> {
     .await
 }
 
+struct MockSubstraitConsumer {
+    extensions: Extensions,
+    state: SessionState,
+}
+
+#[async_trait]
+impl SubstraitConsumer for MockSubstraitConsumer {
+    async fn resolve_table_ref(
+        &self,
+        _: &TableReference,
+    ) -> Result<Option<Arc<dyn TableProvider>>> {
+        not_impl_err!("MockSubstraitConsumer does not implement resolve_table_ref")
+    }
+
+    fn get_extensions(&self) -> &Extensions {
+        &self.extensions
+    }
+
+    fn get_state(&self) -> &SessionState {
+        &self.state
+    }
+
+    async fn consume_extension_leaf(
+        &self,
+        rel: &ExtensionLeafRel,
+    ) -> Result<LogicalPlan> {
+        let Some(ext_detail) = &rel.detail else {
+            return substrait_err!("Unexpected empty detail in ExtensionLeafRel");
+        };
+        let plan = self
+            .state
+            .serializer_registry()
+            .deserialize_logical_plan(&ext_detail.type_url, &ext_detail.value)?;
```

Comment:
One drawback of the current implementation is that it's abusing the `type_url` field in `proto::Any` (which is quite [clearly specified](https://protobuf.dev/reference/java/api-docs/com/google/protobuf/AnyOrBuilder.html#getTypeUrl--). Working with Substrait objects directly would eliminate the need for hacks like encoding node names as Any::type_url.

As a side note, this may sound hypocritical, given that I made it even worse [with my PR](https://github.com/apache/datafusion/pull/13772) by setting it to table names as well. But I only did it for consistency. :)

Reply:
> As a side note, this may sound hypocritical, given that I made it even worse https://github.com/apache/datafusion/pull/13772

I don't think it's hypocritical, you were aiming for consistency and I'm just being much more aggressive with my changes and breaking the API.